### PR TITLE
Untangle Engine Constructor logic

### PR DESCRIPTION
**Pull request by bleskes**
Currently we have a fairly complicated logic in the engine and translog's constructors logic to deal with all the various ways we want to mutate the lucene index and translog we're opening.

We can:
1) Create an empty index
2) Use the lucene but create a new translog
3) Use both
4) Force a new history uuid in all cases.

This leads complicated code flows which makes it harder and harder to make sure we cover all the corner cases. This PR tries to take another approach. Constructing an InternalEngine always opens things as they are and all needed modifications are done by static methods directly on the directory, one at a time.

The PR is still WIP (mostly around documentation) to get initial feedback about what people want think about it. If we like it I can break it up into a translog related PR and an a follow up for the engine.

### CODE REVIEW

**Comment by s1monw** [comment]
the approach makes a lot of sense to me.

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 1]
```diff
@@ -181,22 +177,14 @@ public InternalEngine(EngineConfig engineConfig) {
                 assert translog.getGeneration() != null;
                 this.translog = translog;
                 final IndexCommit startingCommit = getStartingCommitPoint();
-                assert openMode != EngineConfig.OpenMode.OPEN_INDEX_AND_TRANSLOG || startingCommit != null :
-                    "Starting commit should be non-null; mode [" + openMode + "]; startingCommit [" + startingCommit + "]";
+                assert startingCommit != null : "Starting commit should be non-null";
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: Just to remind you that we discussed whether to move startingCommitPoint to EngineDiskUtils.

**Review by dnhatn — COMMENTED** [review]
Looks great to me. I left some minor comments and a reminder.

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 1 above)
Comment (bleskes) [reply server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: I decided to do this as a follow up.

**Comment by bleskes** [comment]
@dnhatn @s1monw I updated this PR and removed the POC label of it. Can you take another look? I think it's ready for full reviews. Thanks!

**Review by dnhatn — APPROVED** [review]
LGTM.

File: server/src/main/java/org/elasticsearch/index/engine/EngineDiskUtils.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+import org.elasticsearch.index.translog.Translog;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/index/engine/EngineDiskUtils.java]: this needs javadocs

File: server/src/main/java/org/elasticsearch/index/engine/EngineDiskUtils.java [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+public final class EngineDiskUtils {
+
+    private EngineDiskUtils() {
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/index/engine/EngineDiskUtils.java]: as an alternative you can mark it as final abstract then you don't need the private ctor

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 4]
```diff
@@ -1349,29 +1316,25 @@ private void innerOpenEngineAndTranslog(final EngineConfig.OpenMode openMode, fi
...
-            final long globalCheckpoint = Translog.readGlobalCheckpoint(translogConfig.getTranslogPath(), translogUUID);
-            replicationTracker.updateGlobalCheckpointOnReplica(globalCheckpoint, "read from translog checkpoint");
-        }
+        // we have to set it before we open an engine and recover from the translog because
+        // acquiring a snapshot from the translog causes a sync which causes the global checkpoint to be pulled in,
+        // and an engine can be forced to close in ctor which also causes the global checkpoint to be pulled in.
+        final String translogUUID = store.readLastCommittedSegmentsInfo().getUserData().get(Translog.TRANSLOG_UUID_KEY);
+        final long globalCheckpoint = Translog.readGlobalCheckpoint(translogConfig.getTranslogPath(), translogUUID);
+        replicationTracker.updateGlobalCheckpointOnReplica(globalCheckpoint, "read from translog checkpoint");
+
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: what are these newlines for here?

**Review by s1monw — APPROVED** [review]
this is a great change! LGTM

File: server/src/main/java/org/elasticsearch/index/engine/EngineDiskUtils.java (hunk 2 above)
Comment (bleskes) [reply server/src/main/java/org/elasticsearch/index/engine/EngineDiskUtils.java]: will add.

File: server/src/main/java/org/elasticsearch/index/engine/EngineDiskUtils.java (hunk 3 above)
Comment (bleskes) [reply server/src/main/java/org/elasticsearch/index/engine/EngineDiskUtils.java]: I don't care. I've seen requests to go both ways. I just want people to be happy. I'll go with abstract. I

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java (hunk 4 above)
Comment (bleskes) [reply server/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: they seperate the flow into units that are related. Something like paragraphs. If you don't like them I'll remove.

**Comment by bleskes** [comment]
Thx @s1monw & @dnhatn

**Comment by bleskes** [comment]
While backporting this to 6.3 I found a flaw in the logic in EngineDiskUtils methods - even if the don't do anything they trigger lucenes deletion policy which messes up with our safe commits. I can disable deletion in these methods but this points a deeper flaw. Since it commits the writer, it also expects the commit it wrote to be used as a starting commit - otherwise it's pointless. For example, it it wants to add a history uuid, it will create a new commit, based on the last one, with the added uuid. We have no guarantee at that point that that commit is actually safe. 

Instead of tweaking these methods, I rather push forward with moving the "remove all unsafe commits" logic to this class. Then I can make sure that the current method only use "safe" commits as a base line.

**Pull request merged**